src/cryptsetup/cryptsetup.c

```diff
@@ -2265,6 +2300,43 @@ static void remove_and_erasep(const char **p) {
                 log_warning_errno(r, "Unable to erase key file '%s', ignoring: %m", *p);
 }
 
+static TokenType determine_token_type(void) {
+        if (arg_tpm2_device || arg_tpm2_device_auto)
+                return TOKEN_TPM2;
+        if (arg_fido2_device || arg_fido2_device_auto)
+                return TOKEN_FIDO2;
+        if (arg_pkcs11_uri || arg_pkcs11_uri_auto)
+                return TOKEN_PKCS11;
+
+        return _TOKEN_TYPE_INVALID;
+}
+
+static int discover_key(const char *key_file, const char *volume, TokenType token_type, struct iovec *ret_key_data) {
+        _cleanup_free_ char *bindname = NULL;
+        const char *token_type_name;
+        int r;
+
+        assert(key_file);
+        assert(volume);
+        assert(ret_key_data);
+
+        bindname = make_bindname(volume, token_type);
+        if (!bindname)
+                return log_oom();
+
+        /* If a key file is not explicitly specified, search for a key in a well defined search path, and load it. */
+        r = find_key_file(key_file, STRV_MAKE("/etc/cryptsetup-keys.d", "/run/cryptsetup-keys.d"), bindname, ret_key_data);
+        if (r <= 0)
+                return r;
+
+        if ((token_type_name = token_type_to_string(token_type)))
```

Comment:
we don't really like that style. please do in two lines. first assign, then check.